geth/transactions/queue/queue.go

```diff
@@ -189,105 +160,77 @@ func (q *TxQueue) Get(id common.QueuedTxID) (*common.QueuedTx, error) {
 	if tx, ok := q.transactions[id]; ok {
 		return tx, nil
 	}
+	return nil, ErrQueuedTxIDNotFound
+}
+
+// LockInprogress returns transcation and locks it as inprogress
+func (q *TxQueue) LockInprogress(id common.QueuedTxID) (*common.QueuedTx, error) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
 
+	if tx, ok := q.transactions[id]; ok {
+		if _, inprogress := q.inprogress[id]; inprogress {
+			return tx, ErrQueuedTxInProgress
+		}
+		q.inprogress[id] = empty{}
+		return tx, nil
+	}
 	return nil, ErrQueuedTxIDNotFound
 }
 
 // Remove removes transaction by transaction identifier
 func (q *TxQueue) Remove(id common.QueuedTxID) {
 	q.mu.Lock()
 	defer q.mu.Unlock()
+	q.remove(id)
+}
 
+func (q *TxQueue) remove(id common.QueuedTxID) {
 	delete(q.transactions, id)
+	delete(q.inprogress, id)
 }
 
-// StartProcessing marks a transaction as in progress. It's thread-safe and
-// prevents from processing the same transaction multiple times.
-func (q *TxQueue) StartProcessing(tx *common.QueuedTx) error {
+// Done removes transaction from queue if no error or error is not transient
+// and notify subscribers
+func (q *TxQueue) Done(id common.QueuedTxID, hash gethcommon.Hash, err error) error {
 	q.mu.Lock()
 	defer q.mu.Unlock()
-
-	if tx.Hash != (gethcommon.Hash{}) || tx.Err != nil {
-		return ErrQueuedTxAlreadyProcessed
+	if tx, ok := q.transactions[id]; !ok {
+		return ErrQueuedTxIDNotFound
+	} else {
+		q.done(tx, hash, err)
 	}
-
-	if tx.InProgress {
-		return ErrQueuedTxInProgress
-	}
-
-	tx.InProgress = true
-
 	return nil
 }
 
-// StopProcessing removes the "InProgress" flag from the transaction.
-func (q *TxQueue) StopProcessing(tx *common.QueuedTx) {
-	q.mu.Lock()
-	defer q.mu.Unlock()
-
-	tx.InProgress = false
+func (q *TxQueue) done(tx *common.QueuedTx, hash gethcommon.Hash, err error) {
+	delete(q.inprogress, tx.ID)
+	tx.Err = err
+	// hash is updated only if err is nil, but transaction is not removed from a queue
+	if err == nil {
+		q.remove(tx.ID)
+		tx.Hash = hash
+		close(tx.Done)
+		return
+	}
+	_, transient := transientErrs[err.Error()]
+	if !transient {
```

Comment:
It can be written as `if _, transient := transientErrs[err.Error()]; !transient {`.